src/Chapter3.hs

```diff
@@ -481,6 +508,45 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
+data MagicalCity = MagicalCity
+  { cityCastle :: Castle,
+    cityBuilding :: CityBuilding,
+    cityHouses :: [House]
+  }
+
+data Castle = None | LoneCastle String | WalledCastle String
+
+data CityBuilding = Church | Library
+
+data House = One | Two | Three | Four
+
+buildCastle :: MagicalCity -> String -> MagicalCity
+buildCastle city name = case cityCastle city of
+  WalledCastle _ -> city {cityCastle = WalledCastle name}
+  _ -> city {cityCastle = LoneCastle name}
+
+buildHouse :: MagicalCity -> House -> MagicalCity
+buildHouse city house = city {cityHouses = house : cityHouses city}
+
+countHouse :: House -> Int
+countHouse house = case house of
+  One -> 1
+  Two -> 2
+  Three -> 3
+  Four -> 4
+
+countPopulation :: MagicalCity -> Int
+countPopulation (MagicalCity _ _ houses) = sum (map countHouse houses)
+
+buildWalls :: MagicalCity -> MagicalCity
+buildWalls city = case cityCastle city of
+  LoneCastle castleName ->
+    if cityPopulation >= 10
+      then city {cityCastle = WalledCastle castleName}
+      else city
+  _ -> city
+  where
+    cityPopulation = countPopulation city
```

Comment:
Absolutely neat solution to the task! Very clear architecture of data types 💯 🏗️

Reply:
Thank you.